src/python/pants/backend/python/tasks/python_binary_create.py

```diff
@@ -126,8 +130,27 @@ def _create_binary(self, binary_tgt, results_dir):
       # Dump everything into the builder's chroot.
       for tgt in source_tgts:
         dump_sources(builder, tgt, self.context.log)
+
+      # Handle locally-built python distribution dependencies.
+      local_dist_req_libs = []
+      built_dists = self.context.products.get_data(BuildLocalPythonDistributions.PYTHON_DISTS)
+      if built_dists:
+        local_dist_req_libs = build_req_libs_provided_by_setup_file(self.context,
+                                                                    built_dists,
+                                                                    self.__class__.__name__)
+      req_tgts = local_dist_req_libs + req_tgts
       dump_requirements(builder, interpreter, req_tgts, self.context.log, binary_tgt.platforms)
 
+      # Dump built python distributions, if any, into builder's chroot.
+      built_dists = self.context.products.get_data(BuildLocalPythonDistributions.PYTHON_DISTS)
+      if built_dists:
+        for dist in built_dists:
+          # Ensure only python_dist dependencies of binary_tgt are added to the output pex.
+          # This protects against the case where a single `./pants binary` command builds two
+          # binary targets that each have their own unique python_dist depencency.
+          if any([tgt.id in dist for tgt in binary_tgt.closure(exclude_scopes=Scopes.COMPILE)]):
+            builder.add_dist_location(dist)
+
```

Comment:
how are you ensuring alignment between the `dump_requirements` provided req->dists vs the dists explicitly added here?

i.e. how does `dump_requirements` know where to resolve custom dists from in the case where they're injected into `req_tgts` above?

Reply:
✅